app/domain/authentication/authn_jenkins/authenticator.rb

```diff
@@ -0,0 +1,111 @@
+      end
+
+      def build_running?(response)
+        body = response.body
+        status_code = response.code
+
+        Rails.logger.debug("Jenkins Build Response: #{body}")
+        if status_code != '200'
+          raise Err::BuildInfoError.new(jenkinsBuildURL, "#{status_code} - #{body}")
+        end
+
+        JSON.parse(body)['building']
+      end
+
+      def jenkins_public_key(response)
+        public_key = response['X-Instance-Identity']
+
+        public_key = "-----BEGIN PUBLIC KEY-----\n#{public_key}\n-----END PUBLIC KEY-----"
+        Rails.logger.debug("Returned public key: #{public_key}")
+        OpenSSL::PKey::RSA.new(public_key)
+      end
+
+      def parse_metadata(username, password)
+        json_body = JSON.parse password
+
+        build_number = json_body["buildNumber"]
+        signature = Base64.decode64(json_body["signature"])
+        job_property_host_prefix = json_body["jobProperty_hostPrefix"]
+        if json_body.key?("jobProperty_hostPrefix")
+          job_name = username.sub("host/#{job_property_host_prefix}/", "")
+        else
+          job_name = username.sub("host/", "")
+        end
+        job_path = job_name.split("/").join("/job/")
+
+        return job_name, job_path, build_number, signature
+      end
+
+      def valid?(input)
+        @account = input.account
+        @service_id = input.service_id
+        @authenticator_name = input.authenticator_name
+        
+        # Get needed secrets to connect into the jenkins API
+        jenkins_username = webservice.variable("jenkinsUsername").secret.value
+        jenkins_password = webservice.variable("jenkinsPassword").secret.value
+        jenkins_url = webservice.variable("jenkinsURL").secret.value
+
+        # Parse the body
+        # e.g {"buildNumber": 5, "signature": "<base64 signature>", "jobProperty_hostPrefix": "myapp"}
+        job_name, job_path, build_number, signature = parse_metadata(input.username, input.password)
+
+        Rails.logger.debug("Job Name: #{job_name} | Build Number: #{build_number} | Signatature: #{signature}")
+
+        # Validate job is running and signature was signed with the jenkins identities private key
+        if build_running?(get_build_info(jenkins_url, job_path, build_number, jenkins_username, jenkins_password))
+          jenkinsPublicKey = jenkins_public_key(http_get_request(jenkins_url, jenkins_username, jenkins_password))
+          message = "#{job_name}-#{build_number}"
+
+          if jenkinsPublicKey.verify(OpenSSL::Digest::SHA256.new, signature, message)
+            return true
+          else
+            Rails.logger.error("AUTHENTICATION FAILED: Data tampered or private-public key mismatch.")
+            raise Err::InvalidSignature
+          end
+        else
+          Rails.logger.error("AUTHENTICATION FAILED: Job '#{job_name} ##{build_number}' is currently not running.")
+          raise Err::RunningJobNotFound "#{job_name} ##{build_number}"
+        end
+      end
```

Comment:
With the above `JenkinsClient`, we could simplify the `valid?` method to something like:
```ruby
def jenkins_client(account, authenticator_name, service_id)
  @jenkins_client ||= begin
    variables = webservice(
      input.account,
      input.authenticator_name,
      input.service_id
    )
    JenkinClient.new(
      variables.variable("jenkinsURL").secret.value,
      variables.variable("jenkinsUsername").secret.value,
      variables.variable("jenkinsPassword").secret.value
    )
  end
end

def valid?(input)
  # Parse the body
  # e.g {"buildNumber": 5, "signature": "<base64 signature>", "jobProperty_hostPrefix": "myapp"}
  job_name, job_path, build_number, signature = parse_metadata(input.username, input.password)

  Rails.logger.debug("Job Name: #{job_name} | Build Number: #{build_number} | Signatature: #{signature}")

  # Validate job is running and signature was signed with the jenkins identities private key
  if build_running?(jenkins_client.build(job_path, build_number))
    public_key = jenkins_client.public_key
    message = "#{job_name}-#{build_number}"

    unless public_key.verify(OpenSSL::Digest::SHA256.new, signature, message)
      Rails.logger.error("AUTHENTICATION FAILED: Data tampered or private-public key mismatch.")
      raise Err::InvalidSignature
    end
  else
    Rails.logger.error("AUTHENTICATION FAILED: Job '#{job_name} ##{build_number}' is currently not running.")
    raise Err::RunningJobNotFound "#{job_name} ##{build_number}"
  end
  true
end
```
We can now mock the `jenkins_client` method, which allows us to easily create the three possible scenarios we can expect from the `valid?` method:
- success
- exception: job not running
- exception: signature mismatch

Reply:
Resolved in commit: https://github.com/AndrewCopeland/conjur/commit/d16c40c290d4f1ac139b101cd089a1635f20db29